src/java/picard/sam/markduplicates/MarkDuplicates.java

```diff
@@ -226,15 +254,28 @@ protected int doWork() {
                     rec.setDuplicateReadFlag(false);
                 }
             }
-            recordInFileIndex++;
 
-            if (!this.REMOVE_DUPLICATES || !rec.getDuplicateReadFlag()) {
-                if (PROGRAM_RECORD_ID != null) {
-                    rec.setAttribute(SAMTag.PG.name(), chainedPgIds.get(rec.getStringAttribute(SAMTag.PG.name())));
+            // Manage the flagging of optical/sequencing duplicates
```

Comment:
@yfarjoun, @eitanbanks : this section is probably the biggest change (along with the change in CLP args), as it now will optionally tag no duplicates, optical only, or all duplicates with a tag.